psycop/common/model_training_v2/hyperparameter_suggester/optuna_hyperparameter_search.py

```diff
@@ -0,0 +1,108 @@
+import copy
+from pathlib import Path
+from typing import Any, Literal
+
+import optuna
+from confection import Config
+from optuna import Study, Trial
+
+from psycop.common.model_training_v2.config.baseline_registry import BaselineRegistry
+from psycop.common.model_training_v2.config.baseline_schema import BaselineSchema
+from psycop.common.model_training_v2.hyperparameter_suggester.hyperparameter_suggester import (
+    SuggesterSpace,
+    suggest_hyperparams_from_cfg,
+)
+from psycop.common.model_training_v2.hyperparameter_suggester.suggesters.base_suggester import (
+    Suggester,
+)
+
+
+class OptunaHyperParameterOptimization:
+    @staticmethod
+    def _validate_configspace(cfg: dict[str, Any]) -> None:
+        has_suggester = (
+            OptunaHyperParameterOptimization._check_if_suggester_in_configspace(cfg=cfg)
+        )
+        if not has_suggester:
+            raise ValueError(
+                "No suggesters found in config. Add one to conduct hyperparameter optimization.",
+            )
+
+    @staticmethod
+    def _check_if_suggester_in_configspace(cfg: dict[str, Any]) -> bool:
+        """Recursively checks if any value in the dictionary or its nested dictionaries
+        is of type SuggesterSpace or Suggester."""
+        for _, value in cfg.items():
+            if isinstance(value, (SuggesterSpace, Suggester)):
+                return True
+            if isinstance(
+                value,
+                dict,
+            ) and OptunaHyperParameterOptimization._check_if_suggester_in_configspace(
+                value,
+            ):
+                return True
+        return False
+
+    @staticmethod
+    def _resolve_only_suggestors(cfg: dict[str, Any]) -> dict[str, Any]:
+        """Resolves only suggesters in a nested config. Suggesters are identified
+        by the presence of the @suggestors key."""
+        cfg_copy = copy.deepcopy(cfg)
+
+        for key, value in cfg_copy.items():
+            if isinstance(value, dict):
+                if "@suggesters" in value:
```

Comment:
Might want to expose this on either the method or self? Ideally taking a list[str] so we can co-locate the suggesters and the implementations (e.g. `estimater_step_suggesters` etc.).

Or we could have it take a regex, like `*._suggesters$`.

In case we:
* Want to reuse this somewhere else, or
* Have to change it in the future

Reply:
Excellent suggestion! Changed to match the regex "^@.*suggester.*", so will resolve if 'suggester' is anywhere in the name of the registry

Reply:
As we just talked about -- agreed, would be extremely neat -- however, not straightforward enough with current architecture. Saving for later